block/submit.go

```diff
@@ -2,200 +2,210 @@ package block
+			submitter.Nudge()
 		}
-		maxTime.Reset(m.Conf.BatchSubmitMaxTime)
-	}
-}
-
-// AccumulatedDataLoop is the main loop for accumulating the produced data size.
-// It is triggered by the ProducedSizeCh channel, which is populated by the block production loop when a new block is produced.
-// It accumulates the size of the produced data and triggers the submission of the batch when the accumulated size is greater than the max size.
-// It also emits a health status event when the submission channel is full.
-func (m *Manager) AccumulatedDataLoop(ctx context.Context, toSubmit chan struct{}) {
-	total := uint64(0)
-
-	/*
-		On node start we want to include the count of any blocks which were produced and not submitted in a previous instance
-	*/
-	currH := m.State.Height()
-	for h := m.LastSubmittedHeight.Load() + 1; h <= currH; h++ {
-		block := m.MustLoadBlock(h)
-		commit := m.MustLoadCommit(h)
-		total += uint64(block.ToProto().Size()) + uint64(commit.ToProto().Size())
-	}
-
-	for {
-		for m.Conf.BlockBatchMaxSizeBytes <= total { // TODO: allow some tolerance for block size (e.g support for BlockBatchMaxSize +- 10%)
-			total -= m.Conf.BlockBatchMaxSizeBytes
+	})
 
+	eg.Go(func() error {
+		// 'submitter': this thread actually creates and submits batches
+		timeLastSubmission := time.Now()
+		for {
 			select {
 			case <-ctx.Done():
-				return
-			case toSubmit <- struct{}{}:
-				m.logger.Info("Enough bytes to build a batch have been accumulated. Sent signal to submit the batch.")
-			default:
-				m.logger.Error("Enough bytes to build a batch have been accumulated, but too many batches are pending submission.	 " +
-					"Pausing block production until a batch submission signal is consumed.")
-
-				evt := &events.DataHealthStatus{Error: fmt.Errorf("submission channel is full: %w", gerrc.ErrResourceExhausted)}
-				uevent.MustPublish(ctx, m.Pubsub, evt, events.HealthStatusList)
-
-				/*
-					Now we block until earlier batches have been submitted. This has the effect of not consuming the producedSizeC,
-					which will stop new block production.
-				*/
-				select {
-				case <-ctx.Done():
-					return
-				case toSubmit <- struct{}{}:
+				return ctx.Err()
+			case <-submitter.C:
+			}
+			pending := pendingBytes.Load()
+			// while there are accumulated blocks, create and submit batches!!
+			for ctx.Err() == nil && (0 < pending && (maxBatchTime < time.Since(timeLastSubmission)) || maxBatchBytes < pending) {
+				nConsumed, err := createAndSubmitBatchGetSizeEstimate(min(pending, maxBatchBytes))
+				if err != nil {
+					return fmt.Errorf("create and submit batch: %w", err)
 				}
-
-				evt = &events.DataHealthStatus{Error: nil}
-				uevent.MustPublish(ctx, m.Pubsub, evt, events.HealthStatusList)
-
-				m.logger.Info("Resumed block production.")
+				timeLastSubmission = time.Now()
+				pending = pendingBytes.Add(^(nConsumed - 1)) // subtract
```

Comment:
shouldn't it be `pending = pendingBytes.Add(^uint64(nConsumed - 1) + 1) // subtract` ?